### PR TITLE
implement condition block, deprecate obsolete inState methods

### LINKED ISSUE #310: DSL for additional conditions
Just an idea that I have mixed feelings about, so dropping my idea here to hear opinions.

```kotlin
enum class MyEnum { A, B, C }
data class MyState(val e : MyEnum)


inState<MyState>(additionalCondition = { it.e == MyEnum.B }) { // means state is MyState AND state.e == B
    ... 
}
```

I am wondering if this could be expressed somehow with DSL constructs. 

For example something like adding a new `condition` DSL block like this:

```kotlin
inState<MyState> {

    condition( {it.e == MyEnum.B} ) {
        onEnter {...}
        on<Action> {...}
        collectWhileInstate { ... }

        condition (...){
              // even subcondition could be supported
         }
     }
   
    
    onEnter {...}
    on<Action> {...}
    collectWhileInstate { ... }

    ... 
}
```

That reads for me a bit nicer compared to the current `inState<S>(additionalCondition={...})`

but obviously has some disadvantages like:
- exhausting `when` functionality alike (could probably be solved with Lint checks or so)
- DSL contains more and more logic that cannot be unit tested anymore individually and reminds me a bit of too much pollated kotest DSL
- ...

Any thoughts?

### PR DESCRIPTION
**Pull request by gabrielittner**
Closes #310, implementation based on what we discussed in #480

The `condition` block is effectively just a second subclass of `BaseBuilderBlock` for which `InStateBuilderBlock` will provide the entry point. As part of this `inStateWithCondition` and `isInState` overloads with `additionalIsInState` parameter have been deprecated.

### CODE REVIEW

File: sample/shared_code/src/commonMain/kotlin/com/freeletics/flowredux/sample/shared/PaginationStateMachine.kt [hunk 1]
```diff
@@ -34,21 +34,15 @@ class InternalPaginationStateMachine(
...
+                condition({ it.canLoadNextPage && it.nextPageLoadingState == NextPageLoadingState.LOADING }) {
+                    onEnter { loadNextPage(it) }
+                }
 
-            inState<ShowContentPaginationState>(additionalIsInState = {
-                it.nextPageLoadingState == NextPageLoadingState.ERROR
-            }) {
-                onEnter { showPaginationErrorFor3SecsThenReset(it) }
-            }
+                condition({ it.nextPageLoadingState == NextPageLoadingState.ERROR }) {
```
Comment (sockeqwe) [inline sample/shared_code/src/commonMain/kotlin/com/freeletics/flowredux/sample/shared/PaginationStateMachine.kt:42]: Reads way better now 👏

**Review by sockeqwe — APPROVED** [review]

**Comment by sockeqwe** [comment]
Should we also update the docummentation while doing the code changes or should we do it as a last task when also  `withIdentity` is implemented?

**Comment by sockeqwe** [comment]
I can take care of the documentation. Will create another branch / living PR where I will update the docs with all the changes we do now....

**Comment by gabrielittner** [comment]
I wanted to do it once all changes are there, thanks for taking care of it

**Pull request merged**